packages/dds/tree/src/simple-tree/core/unhydratedFlexTree.ts

```diff
@@ -150,11 +150,18 @@ export class UnhydratedFlexTreeNode implements UnhydratedFlexTreeNode {
 	public adoptBy(parent: UnhydratedFlexTreeField, index: number): void;
 	public adoptBy(parent: UnhydratedFlexTreeField | undefined, index?: number): void {
 		if (parent !== undefined) {
-			assert(
-				this.location === unparentedLocation,
-				0x98c /* Node may not be adopted if it already has a parent */,
-			);
 			assert(index !== undefined, 0xa08 /* Expected index */);
+			if (this.location !== unparentedLocation) {
+				throw new UsageError("A node may not be in more than one place in the tree");
+			}
+			let n: UnhydratedFlexTreeNode | undefined = parent.parent;
+			while (n !== undefined) {
+				if (n === this) {
+					throw new UsageError("A node may not be recursively inserted into the tree");
+				}
+				// This cast is safe because the parent (if it exists) of an unhydrated flex node is always another unhydrated flex node.
+				n = n.parentField.parent.parent as UnhydratedFlexTreeNode | undefined;
```

Comment:
I love that you documented the safety justification here, but I'd still prefer a runtime check over a cast since its more robust/type safe, especially when considering possible future refactoring:

```suggestion
				// This is safe because the parent (if it exists) of an unhydrated flex node is always another unhydrated flex node.
				const parent = n.parentField.parent.parent;
				assert(parent === undefined || parent instanceof UnhydratedFlexTreeNode, "Unnhydrated node parent should be unhydrated node");
				n = parent;
```

Reply:
Updated it. Also, changed it in the kernel (where I copied this from)